### ISSUE TITLE
Unable to use multiple windows

### ISSUE THREAD

**Issue by bauerdavid** [napari/napari#2977]
## 🐛 Bug

My problem is that when I open multiple napari windows, the context of the window changes completely to the second window. One problem with this, is that when I try to change to 3D mode in the first window, nothing happens, and I get an error:
```
WARNING: Traceback (most recent call last):
  File "C:/Users/Tejfel/AppData/Roaming/JetBrains/PyCharmCE2021.1/scratches/napari_bug_test.py", line 13, in <module>
    napari.run()
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\_qt\qt_event_loop.py", line 376, in run
    app.exec_()
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\utils\action_manager.py", line 55, in <lambda>
    self.command, context
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\utils\action_manager.py", line 43, in call_with_context
    return function(**ctx)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\components\_viewer_key_bindings.py", line 48, in toggle_ndisplay
    viewer.dims.ndisplay = 3
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\utils\events\evented_model.py", line 155, in __setattr__
    getattr(self.events, name)(value=after)  # emit event
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\utils\events\event.py", line 573, in __call__
    self._invoke_callback(cb, event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\utils\events\event.py", line 593, in _invoke_callback
    cb(event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\napari\_vispy\vispy_camera.py", line 138, in _on_ndisplay_change
    self._view.camera = self._3D_camera
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\frozen.py", line 17, in __setattr__
    object.__setattr__(self, key, value)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\scene\widgets\viewbox.py", line 103, in camera
    cam.parent = self.scene
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\scene\node.py", line 200, in parent
    parent._add_child(self)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\scene\node.py", line 214, in _add_child
    self.events.children_change(added=node)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
    cb(event)
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\scene\canvas.py", line 333, in _update_scenegraph
    self.update()
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\scene\canvas.py", line 200, in update
    super(SceneCanvas, self).update()
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\app\canvas.py", line 442, in update
    self._backend._vispy_update()
  File "Y:\BIOMAG\single-cell-raman\src\keras\lib\site-packages\vispy\app\backends\_qt.py", line 427, in _vispy_update
    self.update()
RuntimeError: wrapped C/C++ object of type CanvasBackendDesktop has been deleted
```
You can also see this happening, if you create two windows, close the second one, and try to access `viewer` in the console of the first window. Actually the second window's `viewer` gets returned.
I guess  you didn't create napari to utilize multiple windows simultaneously, so this might be a feature request rather than a bug report, but it would help if I could open a second window.

Actually I can manually set the `viewer` back to the original one programmatically from the console, but this won't help with the ndisplay problem.

Any help would be highly appreciated!
## To Reproduce

1. run code:
```
import napari
from magicgui import magicgui
from napari.layers import Image
import numpy as np

@magicgui(call_button="New window")
def napari_new_viewer():
    viewer2 = napari.Viewer()
    viewer2.add_layer(Image(np.ones((10, 10)), name="image2"))
viewer1 = napari.Viewer()
viewer1.add_layer(Image(np.ones((10, 10)), name="image1"))
viewer1.window.add_dock_widget(napari_new_viewer)
napari.run()
```
2. Click "New window" in the added widget
3. Close second window
4. Run `viewer.layers` in the IPython console
    * "image2" will be returned instead of "image1"
5. Click Toggle ndisplay
    * an error will occure, and it won't change to 3D view

## Expected behavior

For me two options would work well:
a) The context changes when the second window is created, and changes back, when it is closed
b) The app detects that a window (and a viewer) was already created, so it doesn't do anything in the background
In the second window I only need to display the image, segment it, and check it both in 2D and 3D, so a minimal version of napari would be sufficient.

## Environment
napari: 0.4.10
Platform: Windows-10-10.0.19041-SP0
Python: 3.7.5 (tags/v3.7.5:5c02a39a0b, Oct 15 2019, 00:11:34) [MSC v.1916 64 bit (AMD64)]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.19.5
SciPy: 1.4.1
Dask: 2021.06.2
VisPy: 0.6.6

OpenGL:
- GL version: 4.5.0 - Build 25.20.100.6577
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0

Plugins:
- console: 0.0.3
- scikit-image: 0.4.10
- svg: 0.1.5

## Additional context
Actually what I'm trying to do is to allow the user to select a part of the original nD file, and segment it as a separate unit. This would allow the user to visually check it in 3D without loading everything into memory. If it is impossible to use multiple windows simultaneously, than I will try to display the selected part in the original window, and hide the original data, but it would be much nicer and natural to open a second window.

**Comment by jni** [comment]
Thanks for the report @bauerdavid! Actually we do want to be compatible with multiple viewers so this is a big issue. @JoOkuma has done a bunch of work with multiple windows so the `viewer.ndisplay` issue might have more to do with the console. To be clear: do you see the ndisplay issue occurring if you *don't* launch the IPython terminal? (ie if you skip Step 4 in your reproducible example?)

**Comment by jni** [comment]
> so the `viewer.ndisplay` issue might have more to do with the console

... or it might just be a bit of the code that we haven't exercised in this context yet. Either way, it's a bug, and we'll try to fix it ASAP!

**Comment by jni** [comment]
Ah, just saw that this is actually the same issue as #2897. There is more information there and work on a fix has started in #2914. :crossed_fingers:

**Comment by bauerdavid** [comment]
Thanks for the reply! I will watch the ongoing work there.
